Transporter implementation

dbt opened the pull request:
Definitely a WIP

- [x] Copy data from enterprise into transporter
- [x] Insert version data into output schema
- [x] Compute list of ids that have changed and filter updates to only ids that have changed.
- [x] Filter queries with partition ids
- [ ] Test with linked entity sets
- [x] Handle relevant events
- [ ] Do change detection without constantly polling the database

Commit:
Transporter v0. Doesn't know about versions yet

you-leee commented:
Just as a general comment, can you pls provide an example query for linked and non-linked entity set table update? I believe it is really similar to the old one, except the most outer part.
This is an example of the old query for linking entity set:
```
SELECT entity_set_id,
       id,
       entity_key_ids,
       array_agg("im.PersonNickName") FILTER (
                                              WHERE "im.PersonNickName" IS NOT NULL) AS "im.PersonNickName",
       array_agg("nc.PersonSurName") FILTER (
                                             WHERE "nc.PersonSurName" IS NOT NULL) AS "nc.PersonSurName",
       array_agg("nc.PersonHairColorText") FILTER (
                                                   WHERE "nc.PersonHairColorText" IS NOT NULL) AS "nc.PersonHairColorText",
       array_agg("nc.SubjectIdentification") FILTER (
                                                     WHERE "nc.SubjectIdentification" IS NOT NULL) AS "nc.SubjectIdentification",
       array_agg("nc.PersonGivenName") FILTER (
                                               WHERE "nc.PersonGivenName" IS NOT NULL) AS "nc.PersonGivenName",
       array_agg("nc.PersonMiddleName") FILTER (
                                                WHERE "nc.PersonMiddleName" IS NOT NULL) AS "nc.PersonMiddleName",
       array_agg("publicsafety.mugshot") FILTER (
                                                 WHERE "publicsafety.mugshot" IS NOT NULL) AS "publicsafety.mugshot",
       array_agg("nc.PersonBirthDate") FILTER (
                                               WHERE "nc.PersonBirthDate" IS NOT NULL) AS "nc.PersonBirthDate",
       array_agg("nc.PersonEyeColorText") FILTER (
                                                  WHERE "nc.PersonEyeColorText" IS NOT NULL) AS "nc.PersonEyeColorText",
       array_agg("nc.SSN") FILTER (
                                   WHERE "nc.SSN" IS NOT NULL) AS "nc.SSN",
       array_agg("j.SentenceRegisterSexOffenderIndicator") FILTER (
                                                                   WHERE "j.SentenceRegisterSexOffenderIndicator" IS NOT NULL) AS "j.SentenceRegisterSexOffenderIndicator",
       array_agg("nc.PersonSuffix") FILTER (
                                            WHERE "nc.PersonSuffix" IS NOT NULL) AS "nc.PersonSuffix",
       array_agg("nc.PersonHeightMeasure") FILTER (
                                                   WHERE "nc.PersonHeightMeasure" IS NOT NULL) AS "nc.PersonHeightMeasure",
       array_agg("criminaljustice.persontype") FILTER (
                                                       WHERE "criminaljustice.persontype" IS NOT NULL) AS "criminaljustice.persontype",
       array_agg("person.stateidstate") FILTER (
                                                WHERE "person.stateidstate" IS NOT NULL) AS "person.stateidstate",
       array_agg("nc.PersonSex") FILTER (
                                         WHERE "nc.PersonSex" IS NOT NULL) AS "nc.PersonSex",
       array_agg("nc.PersonBirthPlace") FILTER (
                                                WHERE "nc.PersonBirthPlace" IS NOT NULL) AS "nc.PersonBirthPlace",
       array_agg("housing.notes") FILTER (
                                          WHERE "housing.notes" IS NOT NULL) AS "housing.notes",
       array_agg("justice.xref") FILTER (
                                         WHERE "justice.xref" IS NOT NULL) AS "justice.xref",
       array_agg("nc.PersonRace") FILTER (
                                          WHERE "nc.PersonRace" IS NOT NULL) AS "nc.PersonRace",
       array_agg("housing.updatedat") FILTER (
                                              WHERE "housing.updatedat" IS NOT NULL) AS "housing.updatedat",
       array_agg("nc.PersonEthnicity") FILTER (
                                               WHERE "nc.PersonEthnicity" IS NOT NULL) AS "nc.PersonEthnicity"
FROM
  (SELECT entity_set_id,
          id,
          entity_key_ids,
          jsonb_array_elements_text(properties -> 'c270d705-3616-4abc-b16e-f891e264b784')::TEXT AS "im.PersonNickName",
          jsonb_array_elements_text(properties -> '7b038634-a0b4-4ce1-a04f-85d1775937aa')::TEXT AS "nc.PersonSurName",
          jsonb_array_elements_text(properties -> '8293b7f3-d89d-44f5-bec2-6397a4c5af8b')::TEXT AS "nc.PersonHairColorText",
          jsonb_array_elements_text(properties -> '5260cfbd-bfa4-40c1-ade5-cd83cc9f99b2')::TEXT AS "nc.SubjectIdentification",
          jsonb_array_elements_text(properties -> 'e9a0b4dc-5298-47c1-8837-20af172379a5')::TEXT AS "nc.PersonGivenName",
          jsonb_array_elements_text(properties -> 'd0935a7e-efd3-4903-b673-0869ef527dea')::TEXT AS "nc.PersonMiddleName",
          jsonb_array_elements_text(properties -> '45aa6695-a7e7-46b6-96bd-782e6aa9ac13')::TEXT AS "publicsafety.mugshot",
          jsonb_array_elements_text(properties -> '1e6ff0f0-0545-4368-b878-677823459e57')::DATE AS "nc.PersonBirthDate",
          jsonb_array_elements_text(properties -> 'ac37e344-62da-4b50-b608-0618a923a92d')::TEXT AS "nc.PersonEyeColorText",
          jsonb_array_elements_text(properties -> '481f59e4-e146-4963-a837-4f4e514df8b7')::TEXT AS "nc.SSN",
          jsonb_array_elements_text(properties -> 'd9a90e01-9670-46e8-b142-6d0c4871f633')::INTEGER AS "j.SentenceRegisterSexOffenderIndicator",
          jsonb_array_elements_text(properties -> 'd3f3f3de-dc1b-40da-9076-683ddbfeb4d8')::TEXT AS "nc.PersonSuffix",
          jsonb_array_elements_text(properties -> 'f0a6a588-aee7-49a2-8f8e-e5209731da30')::INTEGER AS "nc.PersonHeightMeasure",
          jsonb_array_elements_text(properties -> 'fa00bfdb-98ec-487a-b62f-f6614e4c921b')::TEXT AS "criminaljustice.persontype",
          jsonb_array_elements_text(properties -> '5ea6e8d5-93bb-47cf-b054-9faaeb05fb27')::TEXT AS "person.stateidstate",
          jsonb_array_elements_text(properties -> '6ec154f8-a4a1-4df2-8c57-d98cbac1478e')::TEXT AS "nc.PersonSex",
          jsonb_array_elements_text(properties -> 'cf4598e7-5bbe-49f7-8935-4b1a692f6111')::TEXT AS "nc.PersonBirthPlace",
          jsonb_array_elements_text(properties -> '32eba813-7d20-4be1-bc1a-717f99917a5e')::TEXT AS "housing.notes",
          jsonb_array_elements_text(properties -> 'c7d2c503-651d-483f-8c17-72358bcfc5cc')::TEXT AS "justice.xref",
          jsonb_array_elements_text(properties -> 'f950d05a-f4f2-451b-8c6d-56e78bba8b42')::TEXT AS "nc.PersonRace",
          jsonb_array_elements_text(properties -> '314d2bfd-e50e-4965-b2eb-422742fa265c')::TIMESTAMPTZ AS "housing.updatedat",
          jsonb_array_elements_text(properties -> '1407ac70-ea63-4879-aca4-6722034f0cda')::TEXT AS "nc.PersonEthnicity"
   FROM
     (SELECT entity_set_id,
             id,
             partition,
             array_agg(origin_id) AS entity_key_ids,
             jsonb_object_agg(property_type_id,properties) AS properties
      FROM
        (SELECT entity_set_id,
                id,
                partition,
                property_type_id,
                origin_id,
                COALESCE(jsonb_agg(b_TEXT) FILTER (
                                                   WHERE b_TEXT IS NOT NULL),jsonb_agg(b_UUID) FILTER (
                                                                                                       WHERE b_UUID IS NOT NULL),jsonb_agg(b_SMALLINT) FILTER (
                                                                                                                                                               WHERE b_SMALLINT IS NOT NULL),jsonb_agg(b_INTEGER) FILTER (
WHERE b_INTEGER IS NOT NULL),jsonb_agg(b_BIGINT) FILTER (
                                                         WHERE b_BIGINT IS NOT NULL),jsonb_agg(b_DATE) FILTER (
                                                                                                               WHERE b_DATE IS NOT NULL),jsonb_agg(b_TIMESTAMPTZ) FILTER (
                                                                                                                                                                          WHERE b_TIMESTAMPTZ IS NOT NULL),jsonb_agg(b_DOUBLE) FILTER (
WHERE b_DOUBLE IS NOT NULL),jsonb_agg(b_BOOLEAN) FILTER (
                                                         WHERE b_BOOLEAN IS NOT NULL),jsonb_agg(b_TIMETZ) FILTER (
                                                                                                                  WHERE b_TIMETZ IS NOT NULL),jsonb_agg(n_TEXT) FILTER (
                                                                                                                                                                        WHERE n_TEXT IS NOT NULL),jsonb_agg(n_UUID) FILTER (
WHERE n_UUID IS NOT NULL),jsonb_agg(n_SMALLINT) FILTER (
                                                        WHERE n_SMALLINT IS NOT NULL),jsonb_agg(n_INTEGER) FILTER (
                                                                                                                   WHERE n_INTEGER IS NOT NULL),jsonb_agg(n_BIGINT) FILTER (
                                                                                                                                                                            WHERE n_BIGINT IS NOT NULL),jsonb_agg(n_DATE) FILTER (
WHERE n_DATE IS NOT NULL),jsonb_agg(n_TIMESTAMPTZ) FILTER (
                                                           WHERE n_TIMESTAMPTZ IS NOT NULL),jsonb_agg(n_DOUBLE) FILTER (
                                                                                                                        WHERE n_DOUBLE IS NOT NULL),jsonb_agg(n_BOOLEAN) FILTER (
                                                                                                                                                                                 WHERE n_BOOLEAN IS NOT NULL),jsonb_agg(n_TIMETZ) FILTER (
WHERE n_TIMETZ IS NOT NULL)) AS properties
         FROM DATA
WHERE entity_set_id = ANY(?)
AND partition = ANY(?)
AND VERSION > 0
AND origin_id != '7fffffff-ffff-ffff-7fff-ffffffffffff'
         GROUP BY (entity_set_id,
                   id,
                   partition,
                   property_type_id,
                   origin_id)) entities
      GROUP BY (entity_set_id,
              id,
              partition,
              property_type_id)) AS entity_set_data) AS grouped_data
GROUP BY (entity_set_id,
    id,
    entity_key_ids)
```

you-leee commented:
Sorry for the format, wouldn't allow .sql

[old_mat_query_linking.txt](https://github.com/openlattice/transporter/files/4202025/old_mat_query_linking.txt)

you-leee commented on src/main/kotlin/com/openlattice/transporter/services/TransporterService.kt:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Subscribe
+    fun handleEntityTypeCreated(e: EntityTypeCreatedEvent) {
```
Is the plan to handle all the events, that affect materialized views in transporter? It was just my understanding before, that we keep them in Assembler and use transporter as a service to handle the call (in AssemblerConnectionManager or Assembler)
I mean, in my opinion, both are fine, we just have all that logic, hazelcast processors... already there

dbt replied:
I think it makes sense to put them in the same place, but I don't really care where that is. Transporter doesn't necessarily need to be it's own service, but it's just way easier for me to debug that way right now. We could turn it into a library.

Commit:
get version into tables

also quiet some things and make others louder.

dbt commented:
The queries are going to be pretty similar. I am using property type id instead of FQN for the column name, but we can still use FQN in MVs. I'll upload a query example.

you-leee commented:
Thanks

> The queries are going to be pretty similar. I am using property type id instead of FQN for the column name, but we can still use FQN in MVs. I'll upload a query example.

Thank you! Yeah, I think we need FQNs for user readability.

dbt commented:
Sample query:
[samplequery.txt](https://github.com/openlattice/transporter/files/4202415/samplequery.txt)

you-leee commented:
> Sample query:
> [samplequery.txt](https://github.com/openlattice/transporter/files/4202415/samplequery.txt)

Just a few questions: 
- Shouldn't we include partitions in the query? (Each entity set has some assigned ones)
- How come we are not filtering on `version > <last_updated_version>` and that way we just have to maintain, which table has which version? Because for each update, creat (so any modification) we update the version and set it to current time. And for deletes and soft deletes we have to delete those rows anyway.

makosblade commented on build.gradle:
```diff
@@ -108,6 +111,52 @@ dependencies {
+}
+
+if (T_XMX == 'null' || T_XMX == null || T_XMX == "") {
+    T_XMX = '-Xms1g'
```
should be `-Xmx1g`

makosblade commented on src/main/resources/auth0.yaml:
```diff
@@ -0,0 +1,10 @@
+clientId: "KTzgyxs6KBcJHB872eSMe2cpTHzhxS99"
+clientSecret: "tjAwhBit0deVUor3lcd7YRlHRc_4FJwv_fcLp_ozO2-D2Lh38066IUoGPL4zx1vP"
```
Oooohhhh I hope these aren't real :'(

dbt replied:
We discussed on slack, but no they're not real.

dbt commented:
> * Shouldn't we include partitions in the query? (Each entity set has some assigned ones)

Good call -- I've updated the tracking list.

> * How come we are not filtering on `version > <last_updated_version>` and that way we just have to maintain, which table has which version? Because for each update, creat (so any modification) we update the version and set it to current time. And for deletes and soft deletes we have to delete those rows anyway.

Each table is per entity type, but we're doing updates by entity set, so we can't just globally set the max version per table. And if big imports are happening, at some point we might need to partition the update down to just a slice of ids, so some ids will be ahead of the global max version and it would be nice to have them filtered out.

I'm open to changing this.

you-leee commented:
> > * How come we are not filtering on `version > <last_updated_version>` and that way we just have to maintain, which table has which version? Because for each update, creat (so any modification) we update the version and set it to current time. And for deletes and soft deletes we have to delete those rows anyway.
> 
> Each table is per entity type, but we're doing updates by entity set, so we can't just globally set the max version per table. And if big imports are happening, at some point we might need to partition the update down to just a slice of ids, so some ids will be ahead of the global max version and it would be nice to have them filtered out.
> 
> I'm open to changing this.

Oh, I see, I thought we have tables per entity set.. So I meant to maintain a last updated version for each entity set, but I can see how that would be difficult, if the update contains too many ids.
Btw, just a side info on hard deletions: we have a background task to actually delete those entities called `BackgroundIndexedEntitiesDeletionService`, where the materialized view updates can happen too, otherwise you will loose the info on what entities are hard deleted (or you have to compare ids in mat view table and original table)

dbt commented:
Looking at this deeper, I think it makes sense to skip materializing linked entity sets in the transporter db. We can aggregate them in the end-user views.

Commit:
some refactoring that I want to save before i rewrite some more

Commit:
use data-row as unit of work instead of all propertytypes

Commit:
checkpoint some benchmarked alternatives

katherinebernstein commented on src/main/resources/rhizome.yaml:
```diff
@@ -1,6 +1,34 @@
+    citus: true
+    initialize-indices: true
+    hikari:
+        jdbcUrl: jdbc:postgresql://localhost:5433/openlattice
```
this looks like the citus jdbcUrl -- is the intent that we should be running citus locally to run transporter?

dbt replied:
nope that's just what I'm developing against. this should be the same as the default for conductor, datastore, etc.

Commit:
add another query mode

also switch to doing all the things in an entity type at once

Commit:
narrowing down a solution

Commit:
rounding third and heading for home

updates and queries work

you can't limit an update, so I'll have to CTE that if we want it.
We're already only running one query per ET/PT combo so the sizes
aren't terrible.

dbt commented:
I'm migrating this work into conductor instead of building a different microservice. Most of this is now in openlattice/conductor-client#775
